invert-gradle-plugin/src/main/kotlin/com/squareup/invert/internal/report/sarif/InvertSarifUtils.kt

```diff
@@ -0,0 +1,158 @@
+fun Stat.CodeReferencesStat.CodeReference.toSarifResult(
+    key: StatKey,
+    modulePath: ModulePath?
+): SarifResult = SarifResult(
+    ruleID = key,
+    message = Message(text = code),
+    locations = listOf(
+        Location(
+            physicalLocation = PhysicalLocation(
+                artifactLocation = ArtifactLocation(uri = filePath),
+                region = Region(
+                    startLine = startLine.toLong(),
+                    endLine = endLine.toLong(),
+                    sourceLanguage = determineSourceLanguage(filePath)
+                ),
+                properties = PropertyBag(
+                    extras + mapOf(
+                        "fileType" to filePath.split(".").last()
+                    )
+                )
+            )
+        )
+    ),
+    properties = PropertyBag(
+        mapOf(
+            "owner" to (owner ?: "Unknown"),
+            "module" to modulePath,
+            "uniqueId" to uniqueId,
+        )
+    ),
+    rule = ReportingDescriptorReference(id = key)
+)
+
+/**
+ * Extension function to convert StatsJsReportModel to a map of SARIF rules to their results.
+ */
+@VisibleForTesting
+fun StatsJsReportModel.asSarifRulesAndResults(): Map<ReportingDescriptor, List<SarifResult>> {
+    val rulesAndResults = mutableMapOf<ReportingDescriptor, MutableList<SarifResult>>()
+
+    statsByModule.forEach { (modulePath, statMap) ->
+        statMap.forEach { (statKey, stat) ->
+            // Get or create the rule for this stat
+            val rule = statInfos[statKey]?.asReportingDescriptor() ?: return@forEach
+
+            // Get or create the results list for this rule
+            val results = rulesAndResults.getOrPut(rule) { mutableListOf() }
+
+            // Add results for this stat
+            results.addAll(stat.asSarifResult(modulePath, statKey))
+        }
+    }
+
+    return rulesAndResults
+}
+
+/**
+ * Extension function to convert StatMetadata to SARIF reporting descriptor.
+ */
+@VisibleForTesting
+fun StatMetadata.asReportingDescriptor(shortDescription: String = ""): ReportingDescriptor =
+    ReportingDescriptor(
+        id = key,
+        name = this.title,
+        fullDescription = MultiformatMessageString(markdown = description, text = description),
+        shortDescription = MultiformatMessageString(text = shortDescription),
+        properties = PropertyBag(
+            mapOf(
+                "description" to description,
+                "extras" to extras,
+                "category" to category,
+                "title" to title
+            )
+        )
+    )
+
+@VisibleForTesting
+fun createSarifSchemaFromResults(
+    rule: ReportingDescriptor,
+    results: List<SarifResult>
+): SarifSchema210 {
+    return createSarifSchema(toolName = rule.id, rule = listOf(rule), results = results)
+}
+
+/**
+ * Simple function to determine the source language based on the file path.
+ */
+private fun determineSourceLanguage(filePath: String): String {
```

Comment:
Is this a sarif standard? Is that part of the library or do we need our own manual table?